Add Nessie spec definition for 2.0.0-beta.1 (#6679)
projectnessie/nessie: api/NESSIE-SPEC-2-0.md

# Nessie Specification

Nessie Specifications define the behaviour of Nessie Servers. The same behaviour can be expected to be observable
via all supported API (unless noted otherwise).

[Semver](https://semver.org/spec/v2.0.0.html) principles are followed in bumping the Nessie spec version.
For example, a major behaviour change will result in incrementing the major version number.

The patch version number will be incremented only on minute spec corrections that do not significantly alter the 
server's behaviour.

If a server returns a specific spec version via `NessieConfiguration.getSpecVersion()` it MUST comply with the
specified behaviour (see sections below).

Servers SHOULD return the spec version that defines its behaviour most completely.

> A server SHOULD return `null` from `NessieConfiguration.getSpecVersion()`, when it behaves in a way that is not

Reviewer: Servers don't reply with this field, if the client did not send the corresponding header (so that the server assumes that the client cannot handle the new attributes in NC).

Author: Not quite... `getSpecVersion()` is only affected by the API version, not by headers, AFAIK. The API version restriction should hopefully be clear to users from java annotations and OpenAPI tags :thinking:

compatible with any known spec version.

Servers MAY show behaviours that are not covered by the spec, as long as those behaviours do not contradict the
well-specified behaviours.

> Refer to the Nessie API documentation (under the `model` module) for the meaning of Nessie-specific terms.

Reviewer 2: Beware that this PR has moved this doc one level up (it's not under the `model` module anymore).

Author: but `model` is a sibling of this file now.

Reviewer 2: That was my point... there is no Nessie API documentation under the `model` module anymore, is there? You could change to stg like:

```suggestion
Refer to the [Nessie API documentation](./README.md) for the meaning of Nessie-specific terms.
```

# 2.0.0-beta.1

> * Namespaces are expected to be created before they are referenced. Namespaces may be referenced withing the same 

Reviewer 2: ```suggestion
* Namespaces are expected to be created before they are referenced. Namespaces may be referenced within the same 
```

  commit that creates them. 

> * If a non-existent Namespace is referenced in a commit. The server should fail the corresponding change (commit 

Reviewer 2: ```suggestion
* If a non-existent Namespace is referenced in a commit, the server should fail the corresponding change (commit 
```

  / merge / transplant) with the `NAMESPACE_ABSENT` conflict type.
* `Unchanged` Operations may be submitted in commits, but the server does not persist them. Consequently, `Unchanged`
  operations are not considered during merges and transplants, even when they were part of the original commit.

> * The server MAY consider `Unchanged` operations during the handling of plain commits and MAY raise commit conflicts

Reviewer: ```suggestion
* The server MUST consider `Unchanged` operations during the handling of plain commits and MUST raise commit conflicts
```

Author: fixed

>   if they clash with operations in the commits log since the `expected hash` provided by the client.

Reviewer 2: ```suggestion
  if those operations clash with operations in the commits log since the `expected hash` provided by the client.
```
("they" is ambiguous)

## Related Server Configuration

> Note that setting the `nessie.version.store.advanced.validate-namespaces` configuration property to `false` will 

Reviewer: Should mention the parameters for old _and_ new storage?

Author: fixed

make the server violate the Namespace existence validation rule (above). If the Nessie administration sets this

> property, a custom `NessieConfiguration.json` file must also be provided, which MUST NOT declare `2.0.0-beta.1` as

Reviewer: That's a bit unclear. A user who changed the setting cannot do this.

Author: Shall we check that in code, perhaps?

Author: rephrased
